File: macclean/cleaners/zsh.py
from pathlib import Path
import click
from rich.console import Console
from rich.table import Table
from rich.panel import Panel
from macclean.core.utils import CleanItem, AnalysisResult, format_size, confirm, dir_size

console = Console()


def _dedup_lines(lines: list[str]) -> list[str]:
    seen: set[str] = set()
    result = []
    for line in lines:
        if line not in seen:
            seen.add(line)
            result.append(line)
    return result


def analyze(home: Path | None = None) -> AnalysisResult:
    home = home or Path.home()
    result = AnalysisResult()

    history = home / ".zsh_history"
    if history.exists():
        size = history.stat().st_size
        lines = history.read_text(errors="replace").splitlines()
        deduped = _dedup_lines(lines)
        saved = size - len("\n".join(deduped).encode())
        if saved > 0:
            result.items.append(CleanItem(
                label=f"~/.zsh_history ({len(lines) - len(deduped)} duplicate entries)",
                path=history,
                size_bytes=max(saved, 0),
            ))

    compcache = home / ".zcompcache"
    if compcache.exists():
        result.items.append(CleanItem(
            label="~/.zcompcache (completion cache)",
            path=compcache,
            size_bytes=dir_size(compcache),
        ))

    for dump in home.glob(".zcompdump*"):
        result.items.append(CleanItem(
            label=f"~/{dump.name} (completion dump)",
            path=dump,
            size_bytes=dump.stat().st_size,
        ))

    return result


def clean(result: AnalysisResult, dry_run: bool = False, yes: bool = False) -> None:
    if not result.items:
        console.print("[green]ZSH files are already clean.[/]")
        return

    table = Table(show_header=True, show_lines=True)
    table.add_column("Item")
    table.add_column("Recoverable", justify="right")
    for item in result.items:
        table.add_row(item.label, format_size(item.size_bytes))

    console.print(Panel(table, title="[bold cyan]ZSH Cleanup[/]"))
    console.print(f"  Total recoverable: [bold]{format_size(result.total_bytes)}[/]")

    if dry_run:
        return
    if not yes and not confirm("Clean ZSH history and caches?"):
        return

    import shutil
    home = Path.home()
    history = home / ".zsh_history"
    if history.exists():
        lines = history.read_text(errors="replace").splitlines()
        deduped = _dedup_lines(lines)
        history.write_text("\n".join(deduped) + "\n")
        console.print(f"  [green]✓[/] History deduped ({len(lines) - len(deduped)} entries removed)")

    compcache = home / ".zcompcache"
    if compcache.exists():
        shutil.rmtree(compcache, ignore_errors=True)
        console.print("  [green]✓[/] Removed ~/.zcompcache")

    for dump in home.glob(".zcompdump*"):
        dump.unlink(missing_ok=True)
    console.print("  [green]✓[/] Removed zcompdump files (zsh rebuilds on next open)")


@click.command()
@click.pass_context
def cmd(ctx):
    """Deduplicate zsh history and clean completion caches."""
    result = analyze()
    clean(result, dry_run=ctx.obj["dry_run"], yes=ctx.obj["yes"])

File: src/cleaners/zsh.rs
use std::collections::HashSet;
use std::path::PathBuf;
use anyhow::Result;
use crate::core::{AnalysisResult, Cleaner, CleanItem};
use crate::core::fs::dir_size;
use crate::ui;

pub struct ZshCleaner;

impl Cleaner for ZshCleaner {
    fn name(&self) -> &str { "zsh" }
    fn display_name(&self) -> &str { "ZSH History & Caches" }

    fn analyze(&self) -> Result<AnalysisResult> {
        let home = dirs::home_dir().unwrap_or_else(|| PathBuf::from("/tmp"));
        let mut result = AnalysisResult::default();

        // Check ~/.zsh_history for duplicates
        let history_path = home.join(".zsh_history");
        if history_path.exists() {
            if let Ok(contents) = std::fs::read_to_string(&history_path) {
                let original_size = contents.len() as u64;
                let deduped = dedup_zsh_history(&contents);
                let deduped_size = deduped.len() as u64;
                let saved = original_size.saturating_sub(deduped_size);

                let total_lines = contents.lines().count();
                let deduped_lines = deduped.lines().count();
                let duplicates = total_lines.saturating_sub(deduped_lines);

                if duplicates > 0 {
                    result.items.push(CleanItem {
                        label: format!("{} duplicate history entries", duplicates),
                        path: history_path,
                        size_bytes: saved,
                        removable: true,
                    });
                }
            }
        }

        // Check ~/.zcompcache
        let zcompcache = home.join(".zcompcache");
        if zcompcache.exists() {
            let size = dir_size(&zcompcache);
            result.add("ZSH completion cache (~/.zcompcache)", zcompcache, size);
        }

        // Check ~/.zcompdump* files
        let home_entries = match std::fs::read_dir(&home) {
            Ok(e) => e,
            Err(_) => return Ok(result),
        };
        for entry in home_entries.filter_map(|e| e.ok()) {
            let name = entry.file_name().to_string_lossy().to_string();
            if name.starts_with(".zcompdump") {
                let path = entry.path();
                let size = path.metadata().map(|m| m.len()).unwrap_or(0);
                result.add(format!("ZSH completion dump ({})", name), path, size);
            }
        }

        Ok(result)
    }

    fn clean(&self, result: &AnalysisResult, dry_run: bool, yes: bool) -> Result<()> {
        if result.items.is_empty() {
            println!("Nothing to clean in ZSH history/caches.");
            return Ok(());
        }
        ui::print_analysis("ZSH History & Caches", &result.items);
        if dry_run { return Ok(()); }
        if !yes && !ui::confirm("Clean ZSH history and caches?", false)? { return Ok(()); }

        let home = dirs::home_dir().unwrap_or_else(|| PathBuf::from("/tmp"));

        for item in &result.items {
            // History dedup
            if item.path == home.join(".zsh_history") {
                match dedup_and_write_history(&item.path) {
                    Ok(_) => ui::print_ok("Deduplicated ZSH history"),
                    Err(e) => ui::print_warn(&format!("ZSH history: {}", e)),
                }
                continue;
            }

            // Completion cache dir
            if item.path.is_dir() {
                match std::fs::remove_dir_all(&item.path) {
                    Ok(_) => ui::print_ok(&format!("Removed {}", item.label)),
                    Err(e) => ui::print_warn(&format!("{}: {}", item.label, e)),
                }
            } else {
                // Completion dump files
                match std::fs::remove_file(&item.path) {
                    Ok(_) => ui::print_ok(&format!("Removed {}", item.label)),
                    Err(e) => ui::print_warn(&format!("{}: {}", item.label, e)),
                }
            }
        }
        Ok(())
    }
}

/// Deduplicate zsh history lines, preserving order (last occurrence wins for extended history).
/// Handles both plain lines and extended_history format (": timestamp:elapsed;command").
fn dedup_zsh_history(contents: &str) -> String {
    let mut seen: HashSet<String> = HashSet::new();

    // Walk lines in reverse to keep the last occurrence of each command
    let lines: Vec<&str> = contents.lines().collect();
    let mut result_rev: Vec<&str> = Vec::new();

    for line in lines.iter().rev() {
        // Extract the command part for dedup key
        let key = if line.starts_with(": ") {
            // extended_history format: ": timestamp:elapsed;command"
            line.splitn(2, ';').nth(1).unwrap_or(line).trim()
        } else {
            line.trim()
        };

        if seen.insert(key.to_string()) {
            result_rev.push(line);
        }
    }

    result_rev.reverse();
    let kept = result_rev;

    let mut out = kept.join("\n");
    if !out.is_empty() && contents.ends_with('\n') {
        out.push('\n');
    }
    out
}

fn dedup_and_write_history(path: &std::path::Path) -> Result<()> {
    let contents = std::fs::read_to_string(path)?;
    let deduped = dedup_zsh_history(&contents);
    std::fs::write(path, deduped.as_bytes())?;
    Ok(())
}
